openmc/material.py

```diff
@@ -0,0 +1,1109 @@
+        for nuc, atoms_per_cc in self.get_nuclide_atom_densities().values():
+            Z = openmc.data.zam(nuc)[0]
+            if Z >= 90:
+                density += 1e24 * atoms_per_cc * openmc.data.atomic_mass(nuc) \
+                           / openmc.data.AVOGADRO
+        return density*self.volume
+
+    @classmethod
+    def from_hdf5(cls, group):
+        """Create material from HDF5 group
+
+        Parameters
+        ----------
+        group : h5py.Group
+            Group in HDF5 file
+
+        Returns
+        -------
+        openmc.Material
+            Material instance
+
+        """
+        mat_id = int(group.name.split('/')[-1].lstrip('material '))
+
+        name = group['name'][()].decode() if 'name' in group else ''
+        density = group['atom_density'][()]
+        if 'nuclide_densities' in group:
+            nuc_densities = group['nuclide_densities'][()]
+
+        # Create the Material
+        material = cls(mat_id, name)
+        material.depletable = bool(group.attrs['depletable'])
+        if 'volume' in group.attrs:
+            material.volume = group.attrs['volume']
+        if "temperature" in group.attrs:
+            material.temperature = group.attrs["temperature"]
+
+        # Read the names of the S(a,b) tables for this Material and add them
+        if 'sab_names' in group:
+            sab_tables = group['sab_names'][()]
+            for sab_table in sab_tables:
+                name = sab_table.decode()
+                material.add_s_alpha_beta(name)
+
+        # Set the Material's density to atom/b-cm as used by OpenMC
+        material.set_density(density=density, units='atom/b-cm')
+
+        if 'nuclides' in group:
+            nuclides = group['nuclides'][()]
+            # Add all nuclides to the Material
+            for fullname, density in zip(nuclides, nuc_densities):
+                name = fullname.decode().strip()
+                material.add_nuclide(name, percent=density, percent_type='ao')
+        if 'macroscopics' in group:
+            macroscopics = group['macroscopics'][()]
+            # Add all macroscopics to the Material
+            for fullname in macroscopics:
+                name = fullname.decode().strip()
+                material.add_macroscopic(name)
+
+        return material
+
+    def add_volume_information(self, volume_calc):
+        """Add volume information to a material.
+
+        Parameters
+        ----------
+        volume_calc : openmc.VolumeCalculation
+            Results from a stochastic volume calculation
+
+        """
+        if volume_calc.domain_type == 'material':
+            if self.id in volume_calc.volumes:
+                self._volume = volume_calc.volumes[self.id].n
+                self._atoms = volume_calc.atoms[self.id]
+            else:
+                raise ValueError('No volume information found for this material.')
+        else:
+            raise ValueError('No volume information found for this material.')
```

Comment:
Consider including the `domain_type` and/or repr/str/id in the error messages here so callers can more easily debug their code.